R/lassnags.r

```diff
@@ -0,0 +1,246 @@
+#' @return Nothing, the point cloud is updated by reference. A new field is added to the las file (\code{SnagCls}),
+#' for each point with the corresponding classificaiton: 0) live tree, 1) general snag, 2) small snag, 3) live 
+#' crown edge snag, and 4) high canopy cover snag.
+#'
+#' @section Wing et al. 2016:
+#' This is an automated filtering algorithm that utilizes three dimensional neighborhood lidar point-based
+#' intensity and density statistics to remove lidar points associated with live trees and retain lidar points
+#' associated with snags devleoped by Wing et al (2015 - see references). 
+#' Note that this algorithm strictly performs a classifciation based on user input while the original publicaiton's 
+#' methods also included a segmentation step and some pre- (filtering for first and single returns only) and post-
+#' process (filtering for only the snag classified points prior to segmentation) tasks whihc are now expected to 
+#' be preformed by the user. Also, this implemntation may have some differences compared with the original method
+#' due to potential mis-interpretation of the Wing et al. manuscript, specifically Table 2 where they present four
+#' groups of conditional assessments with their required neighborhood point density and average BBPR values (BBPR 
+#' = branch and bole point ratio; PDR = point density requirement). 
+#' 
+#' \describe{
+#' \item{\code{neigh_radii}}{A vector of three radii used in quantifying local-area centered neighborhoods. See 
+#' reference page 171 and Figure 4. Defaults are 1.5, 1, and 2 for the sphere, small cylinder and large cylinder neighborhoods,
+#' respectively}
+#' \item{\code{low_int_thrsh}}{The lower intensity threshold filtering value. See reference page 171. Default is 50}
+#' \item{\code{uppr_int_thrsh}}{The upper intensity threshold filtering value. See reference page 171. Default is 170}
+#' \item{\code{pt_den_req}}{Point density requirement based on plot-level point density defined classes. See reference
+#' page 172. Default is 3}
+#' \item{\code{bbpr_thresholds}}{A 3x4 matrix providing the four average BBPR values for each of the three neighborhood
+#' (sphere, small cylinder and large cylinder neighborhoods) to be used in for conditional assessments and classificaiton
+#' into the following four snag classes: 1) general snag, 2) small snag, 3) vlive crown edge snag, and 4) high canopy 
+#' cover snag. See reference page 172 and Table 2. This matirx must be provided by the user.}
+#' The current implementation is known to use a large amount of memory for storing the N x k integer matrix returning the
+#' near neighbour indices for each point in the point cloud. Improvements are possible in future package versions.}
+#' 
+#' @examples
+#' LASfile <- system.file("extdata", "MixedConifer.laz", package="lidR")
+#' las = readLAS(LASfile, select = "xyzi")
+#' lasnormalize(las, grid_terrain(las, method = "knnidw", k = 10L))
+#' las %<>% lasfilter(Z>1.37, ReturnNumber==1)
```

Comment:
Use streaming filter in readLAS. It's fine since we don't normalize anymore

Reply:
Holy cow - that's awesome and I completely missed it!  ;)

Reply:
You're not the only one. This is why I try to emphasis it in the doc